### ISSUE TITLE
Cross-component dependencies

### ISSUE THREAD

**Issue by tauren** [react-boilerplate/react-boilerplate#1264]
Apologies in advance for being a little long-winded. **tldr**: Aren't all the cross-component dependencies in this boilerplate encouraging the development of monolithic apps? Is that even something I should be concerned with?

---

I've been using this boilerplate for a few weeks now and for the most part am pleased with it. Thanks to everyone for all the effort put into it! One reason I selected it was that I very much agree with organizing files into component folders. This is so much better than having a folder for each type of file (actions folder, reducers folder, images folder, etc.)

However, now that I've spent some time with it and have a greater understanding of the structure, I'm finding it hard to eliminate cross-component dependencies. This concerns me, as I've spend the last few years moving away from tightly-coupled monolithic designs toward isolated, modular, microservice architectures.

Is it just me, or are things at odds with one another with this boilerplate? We want our components to be isolated, but still provide a way to interact with the rest of the application. We want to avoid code duplication, so we write action creators, selectors, sagas, and event name constants. But often, more than one component will need to use those actions, selectors, sagas, and constants. So we end up with a web of imported modules that becomes increasingly harder to keep track of as the application grows. And then there is the question of where to put things like authentication reducers/selectors/sagas/etc that are needed in multiple components. 

Here are some imports in a recent version of `react-boilerplate` that reference files outside their scope, either from a parent or a sibling:

```js
// from app/containers/HomePage/index.js
import { loadRepos } from '../App/actions';
import { selectRepos, selectLoading, selectError } from 'containers/App/selectors';

// from app/containers/LanguageProvider/reducer.js
import { DEFAULT_LOCALE } from '../App/constants';

// from app/containers/LocaleToggle/index.js
import { appLocales } from '../../i18n';
import { changeLocale } from '../LanguageProvider/actions';
import { selectLocale } from '../LanguageProvider/selectors';
```

I realize this is a simplistic example, but in my own much larger app based on this boilerplate, I'm finding it increasingly difficult to isolate features. Sagas in one component are needing to access selectors, actions, and constants in other components or from the `app` level. It just seems *dirty* to me. I worry about code-splitting and reuse. But I don't have a better solution.

I would love to hear from @mxstbr and other contributors about these design decisions. Having read through many of the past issues, I know there has been lots of talk about the organization and design of this boilerplate. I'm sure there are good reasons for many of the choices.

* Is the design simply a symptom of the react and redux architecture?
* If not, was it an intentional design decision and what was the reasoning?
* If this is not the desired design, do you simply need more time to better separate concerns?
* Is the monolithic nature of apps built in this manner not concerning to anyone else?
* Or have these cross-dependencies just not yet been considered?
* How do these cross-dependencies affect code-splitting?
* Without better isolation, how can these components be reused in other apps, especially if the state shape is not the same?

For a little background, I recently built a large enterprise real-time application using web components (via Polymer). I'm not trying to compare React/Redux with it in the least, but the design pretty much enforced isolation. It was important to build components that were portable and reusable, so I'm struggling to accept some of the patterns I'm seeing here.

Note that this is my first large-scale React project, but I have years of full-stack javascript experience. So perhaps I haven't yet learned the right way to do things in react/redux. Or maybe this is just the norm for react/redux development and I need more time to get used to it.

**Comment by kopax** [comment]
I have probably less experience with it than you but since we started to work using a clean react boilerplate I can see that the more it grows the more I see cross-imports. 

We are rewriting bootstrap 4  using 100% styled components from a-Z incl doc.

To get the full features for my components: 

- Components that work with theming
- Components that has default theming values
- Components that can also work without theming
- Components that can also work for react/native
- Being able to share it quickly to the opensource community

Problem is it's a lot of work to figure out how to remove properly theses imports and create packages quickly like in Java. I feel it's almost impossible in some cases.

I think it's an interesting question, thanks for asking it.

**Comment by tinybug** [comment]
It's really not a good practice access selectors, actions, and constants in other components. To simplify and make the components clean, I would like to put all the share selector, actions, and constants in the global scope(the app level). I know it will make the app level complicate, but it's easier to maintain the whole project.

**Comment by samit4me** [comment]
@kopax Is bootstrap going to be using the [`styled-components`](https://github.com/styled-components/styled-components)?

@tauren Great question, it appears to be mostly related to Redux! I personally believe Redux is popular because it is abstract. This is good and bad, GOOD as it enables flexibility so you can bend it to your needs, BAD bad as it doesn't really give you an opinionated way on how to do something.

In terms of [code structure](http://redux.js.org/docs/faq/CodeStructure.html#code-structure), this boilerplate follows the "Domain-style", which works well when each folder is self-contained.

> But often, more than one component will need to use those actions, selectors, sagas, and constants. So we end up with a web of imported modules that becomes increasingly harder to keep track of as the application grows.

That is correct, for a lot of apps, there will be shared state, actions, and selectors. How to achieve this all comes down to personal preferences and what suits your specific needs (there is no one-size fits all). One approach I've found that works fairly well is to place all shared Redux "stuff" under the App container (e.g. `containers/App/` in this boilerplate), but there are many ways to achieve this.

There is a [great article from Tal Kol on Medium](https://hackernoon.com/redux-step-by-step-a-simple-and-robust-workflow-for-real-life-apps-1fdf7df46092#.7a2laggew) discussing the topic of structuring. Another fantastic resource for all things React / Redux is [React/Redux Links](https://github.com/markerikson/react-redux-links#reactredux-links) by Mark Erikson.

Not sure anything I've said will actually help, but let us know if you find something useful :)

**Comment by kopax** [comment]
@samit4me, yes.
You have components store on a side under `components/bootstrap/{atoms/molecules/organisms}`
Beside you also have mixins/utilities available in `styled/helpers/bootstrap/{mixins/utilities}` and configuration `styled/themes/bootstrap/{atoms/molecules/organisms}` with easy override by keeping default values all the time. Works pretty well. We are redoing the bootstrap documentation, I was thinking to share this work if the community can help to maintain and improve.
We didn't kept the bootstrap.js, it's all [styled-components](https://github.com/styled-components/styled-components).

**Comment by carlos-peru** [comment]
@taurent, I think the purpose is to have thin containers (aka container components). The container components should be reusing components (aka presentational components). In a larger feature, any global state should be passed down as a prop from the container to the component, so the components can still be reusable.

The piece of code from the example could be improved by letting LanguageProvider have all the locale related values and states. DEFAULT_LOCALE should not be in App/constants but in LanguageProvider/constants.

Then the HomePage should have, in that example case, all the repos within the its own HomePage folder. I think in this example case was done like that just to show how you can access the parent container but it was just not a good example.

That way, the containers should only import global values that are used to handle global states like routes, locale, session (auth, user, however you want to call it).

I'm not familiar enough to the bootstrap work done by @kopax, but maybe trying to have a clear distinction between local and global states can help (maybe themes related stuff can be kept in a global state).

**Comment by samit4me** [comment]
Here is a classic example of how there is no correct way, the docs refer to a [medium article](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0#.bsc66fj43) about how the author finds it useful to separate Presentational and Container components and then regrets ever writing the article in this [tweet](https://twitter.com/dan_abramov/status/802569801906475008). I think the takeaway from this is, experiment and try different ways and find what works best for your use case.

**Comment by samit4me** [comment]
Closing the issue as there is no clear action to be taken, but feel free to continue the discussion.

**Comment by hunterwilliams** [comment]
Not sure whether to leave a comment here or bring up a new issue.

In development, on multiple projects we've noticed that we end up with DEFAULT_LOCALE in both App/constants and LanguageProvider/constants. Additionally available locales ends up being placed in both LanguageProvider/constants and i18n. 

I believe ideally it gets moved into a single location. Previously, in a later stage project we removed it from App/constants thinking this was fine as it wasn't used anywhere internally, but neglected to check the scripts folder. This was quickly found once extract-intl breaks, but it seems weird to need DEFAULT_LOCALE in two places therefore possibly with a conflicting value. 

Thoughts?
